shadowsocks/cipher.go

```diff
@@ -0,0 +1,132 @@
+// Copyright 2020 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shadowsocks
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/sha1"
+	"fmt"
+	"io"
+	"strings"
+
+	"golang.org/x/crypto/chacha20poly1305"
+	"golang.org/x/crypto/hkdf"
+)
+
+// SuportedCipherNames lists the names of the AEAD ciphers that are supported.
+var SuportedCipherNames []string
+
+func init() {
+	SuportedCipherNames = make([]string, len(supportedAEADs))
+	for i, spec := range supportedAEADs {
+		SuportedCipherNames[i] = spec.name
+	}
+}
+
+type aeadSpec struct {
+	name        string
+	newInstance func(key []byte) (cipher.AEAD, error)
+	keySize     int
+	tagSize     int
+}
+
+// List of supported AEAD ciphers, as specified at https://shadowsocks.org/en/spec/AEAD-Ciphers.html
+var supportedAEADs = []aeadSpec{
+	newAeadSpec("chacha20-ietf-poly1305", chacha20poly1305.New, chacha20poly1305.KeySize),
+	newAeadSpec("aes-256-gcm", newAesGCM, 32),
+	newAeadSpec("aes-192-gcm", newAesGCM, 24),
+	newAeadSpec("aes-128-gcm", newAesGCM, 16),
+}
+
+func newAeadSpec(name string, newInstance func(key []byte) (cipher.AEAD, error), keySize int) aeadSpec {
+	dummyAead, err := newInstance(make([]byte, keySize))
+	if err != nil {
+		panic(fmt.Sprintf("Failed to initialize AEAD %v", name))
+	}
+	return aeadSpec{name, newInstance, keySize, dummyAead.Overhead()}
+}
+
+func newAesGCM(key []byte) (cipher.AEAD, error) {
+	blk, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+	return cipher.NewGCM(blk)
+}
+
+// Cipher encapsulates a Shadowsocks AEAD spec and a secret
+type Cipher struct {
+	aead   aeadSpec
+	secret []byte
+}
+
+// SaltSize is the size of the salt for this Cipher
+func (c *Cipher) SaltSize() int {
+	return c.aead.keySize
+}
+
+// TagSize is the size of the AEAD tag for this Cipher
+func (c *Cipher) TagSize() int {
+	return c.aead.tagSize
+}
+
+// NewAEAD creates the AEAD for this cipher
+func (c *Cipher) NewAEAD(salt []byte) (cipher.AEAD, error) {
+	sessionKey := make([]byte, c.aead.keySize)
+	r := hkdf.New(sha1.New, c.secret, salt, []byte("ss-subkey"))
+	if _, err := io.ReadFull(r, sessionKey); err != nil {
+		return nil, err
+	}
+	return c.aead.newInstance(sessionKey)
+}
+
+func getAEADSpec(name string) (*aeadSpec, error) {
```

Comment:
I think `getAEADSpec` should probably be moved closer to `newAeadSpec`.  Also, the capitalization should match.

Reply:
Done